### PR TITLE
Wip/hughsie/block user agents

### PR DESCRIPTION
**Pull request by hughsie**

### CODE REVIEW

File: app/util.py [hunk 1]
```diff
@@ -16,6 +16,39 @@
 from gi.repository import GCab
 from gi.repository import GLib
 
+def _vercmp(ver1, ver2):
```
Comment (superm1) [inline app/util.py]: You've already got version comparing from appstream glib no?  Why not use that?

File: app/views.py [hunk 2]
```diff
@@ -11,17 +11,40 @@
...
+from .util import _error_internal, _error_permission_denied, _vercmp
+
+def _user_agent_safe_for_requirement(user_agent):
+
+    # very early versions of fwupd used 'fwupdmgr' as the user agent
+    if user_agent == 'fwupdmgr':
+        return False
+
+    # gnome-software/3.26.5 (Linux x86_64 4.14.0) fwupd/1.0.4
+    sections = user_agent.split(' ')
```
Comment (superm1) [inline app/views.py]: I think you can combine this and the bottom for loop for an optimization.

```
for chunk in sections:
    toks = chunk.split('/')
    if len(toks) == 2:
        if toks[0] == 'fwupd':
            return _vercmp('0.8.0', toks[1]) >= 0
        else if toks[0] == 'gnome-software':
            return _vercmp('3.26.0', toks[1]) >= 0
```

File: uploadedfile_test.py [hunk 3]
```diff
@@ -75,6 +75,34 @@ def archive_sign(self, arc, firmware_cff):
             return
         _archive_add(arc, detached_fn, 'signed')
 
+class VersionTestCase(unittest.TestCase):
```
Comment (superm1) [inline uploadedfile_test.py]: If you can end up using the appstream glib version compare you can probably drop this test case.

File: app/views.py (hunk 2 above)
Comment (hughsie) [reply app/views.py]: Nahh, we can't unroll it as we want to match `fwupd/1.0.4` before `gnome-software/3.26.5` -- i.e. it's the last token. I suppose we could reverse the sections although that's a little magic.

File: app/util.py (hunk 1 above)
Comment (hughsie) [reply app/util.py]: Ohh, and this is why you shouldn't code at 10pm. I'll use that indeed.

File: uploadedfile_test.py (hunk 3 above)
Comment (hughsie) [reply uploadedfile_test.py]: Fixed.

**Comment by hughsie** [comment]
I'll merge this as I've got other things depending on this branch. Thanks all for comments so far. If Ubuntu could merge the 16.04 that'd make me feel warm and fuzzy.

**Pull request merged**